File: OTFS_DP/MatlabBaseCode/OTFS_Book_Code/Transmiter.m
%number of Doppler bins (time slots)
N= 16;

%number of delay bins(subcarriers)
M=64;

% normalized DFT matrix
Fn = dftmtx(N);
Fn = Fn/norm(Fn);

%subcarrier spacing
delta_f = 15e3;

%block duration
T = 1/delta_f;

%carrier frequency 
fc=4e9;

% speed of light
c = 299792458;

%OTFS grid delay and Doppler resolution

delay_resolution = 1/(M*delta_f);
Doppler_resolution = 1/(N*T);

%% modulation
%modulation size
mod_size = 4;
%Number of information symboles in one frame
N_syms_per_frame = N*M;
% number of informaton bits in one frame
N_bits_per_frame = N*M*log2(mod_size);
%generate random bits
tx_info_bits = randi([0,1],N_bits_per_frame,1);

%QAM modulation 
tx_info_bits = qammod(tx_info_bits,mod_size,'gray','InputType','bit');

%Generate the MxN OTFS delay_Doppler frame
X = reshape(tx_info_bits,M,N);
x = reshape(X.',N*M,1);%X.' the dot is to dont modify complex part

Im = eye(M);

%row-column permutation matrix
P=zeros(N*M,N*M);
for j=1:N
    for i=1:M
        E=zeros(M,N);
        E(i,j)=1;
        P((j-1)*M+1:j*M,(i-1)*N+1:i*N)=E;
    end
end



% Method 1 (Eqs. (4.19) and (4.20))
X_tilda=X*Fn';
s=reshape(X_tilda,1,N*M);
% Method 2 (Eq. (4.35))
s1=P*kron(Im,Fn')*x;
% Method 3 (Eq. (4.35))s
s2=kron(Fn',Im)*P*x;

File: OTFS_DP/PythonBaseCode/OTFS/BookTranslation/Transmiter.py
from mimetypes import init
from random import randint
import numpy as np
from numpy.linalg import norm 
from math import log2
from QAM_mod import QPSK
from Matrix_tools import Permutation_matrix

class Transmiter():
    def __init__(self,N=16,M=64,delta_f=15e3,QAM_symbols=4):
        self.N = N
        self.M = M
        self.P = None
        # normalized DFT matrix
        self.Fn = np.fft.fft(np.eye(N))
        self.Fn = self.Fn/norm(self.Fn)
        #subcarrier spacing
        self.delta_f = delta_f
        #block duration
        self.T = 1/self.delta_f
        #carrier frequency 
        self.fc=4e9
        #OTFS grid delay and Doppler resolution
        self.delay_resolution   = 1/(self.M*self.delta_f)
        self.Doppler_resolution = 1/(self.N*self.T)
        #QAM modulation symbols
        self.mod_size = QAM_symbols
        #Symbols per frame
        self.N_syms_per_frame = self.N*self.M
        #Number of informaton bits in one frame
        N_bits_per_frame = self.N*self.M*log2(self.mod_size)
        #QAM modulation
        self.tx_info_bits = None
        if(self.mod_size == 4):
            self.tx_info_bits = QPSK(N_bits_per_frame)
        else:
            #TODO Generik constelation or Higher QAM
            pass
        self.X = np.reshape(self.tx_info_bits,(M,N))
        #vectorized version
        self.vect_x = np.reshape(self.X.T,(N*M,1))
        
        
    #Delay Doppler transformed 
    #to delay-time with Inveze ZAK transform
    def Inverse_ZAK_method(self):
        X_tilda=self.X*self.Fn
        s = np.reshape(X_tilda,(1,self.N*self.M))
        return s
    #Permutaton matrix implement
    def Permutation_method(self):
        self.P  = Permutation_matrix(self.N,self.M)
        Im = np.eye(self.M)
        s  = self.P@np.kron(Im,self.Fn.T)@self.vect_x
        return s